add GitHub actions workflow file

andy5995 opened the pull request:
for #2472

andy5995 commented:
I'm setting this up on the master branch of my fork because I don't think anything would run otherwise.

Here's the current build output (which is failing)

https://github.com/andy5995/appimage.github.io/actions

andy5995 commented on code/worker.sh:
```diff
@@ -3,8 +3,8 @@
 
-if [ x"$TRAVIS_PULL_REQUEST" == xfalse ] ; then
-  git checkout "$TRAVIS_BRANCH"
+if [ x"$GITHUB_ACTIONS" == xfalse ] ; then
```
I'm not too sure about these 2 vars and their Travis/Actions equivalent. Suggestions?

JulianGro commented on .github/workflows/test.yml:
```diff
@@ -0,0 +1,41 @@
+
+jobs:
+  build:
+    runs-on: ubuntu-latest
```
```suggestion
    runs-on: ubuntu-20.04
```
I would suggest using ubuntu-20.04 instead, so you don't have a later update to Ubuntu 22.04 break something randomly.

JulianGro commented on .github/workflows/test.yml:
```diff
@@ -0,0 +1,41 @@
+    - uses: actions/checkout@v2
+    - name: Install dependencies
+      run: |
+        sudo apt-get -qq -y install imagemagick libasound2-dev pulseaudio-utils alsa-utils alsa-oss libjack0 desktop-file-utils xmlstarlet xterm xvfb icewm x11-utils x11-apps netpbm xdotool libgl1-mesa-dri libgl1-mesa-dev mesa-utils libosmesa6 libsdl1.2-dev libsdl2-2.0-0 fonts-wqy-microhei libfile-mimeinfo-perl > /dev/null # appstream # TODO: Cache me!
```
```suggestion
        sudo apt-get -qq -y install imagemagick libasound2-dev pulseaudio-utils alsa-utils alsa-oss libjack0 desktop-file-utils xmlstarlet xterm xvfb icewm x11-utils x11-apps netpbm xdotool libgl1-mesa-dri libgl1-mesa-dev mesa-utils libosmesa6 libsdl1.2-dev libsdl2-2.0-0 fonts-wqy-microhei libfile-mimeinfo-perl # appstream # TODO: Cache me!
```
Are you sure you want to output everything to `/dev/null`? What if there is an issue with installing the packages? I have seen rare problems with GitHub Actions runners having unreliable network connections and thous failing to download packages for example.

JulianGro commented on code/worker.sh:
```diff
@@ -533,7 +535,7 @@ echo "==========================================="
 # If this a PR, then just check whether the files have generated
 # See https://github.com/AppImage/appimage.github.io/issues/476 for more information
-if [ "$TRAVIS_PULL_REQUEST" != "false" ]; then 
+if [ "$TRAVIS_PULL_REQUEST" != "false" ]; then
```
```suggestion
if [ "$github.event.pull_request" != "false" ]; then
```
Not 100% sure about this.

andy5995 replied:
I think `"$github.event.pull_request"` is only valid in the .yml. Though according to https://stackoverflow.com/questions/61886993/in-github-actions-how-to-get-the-type-of-a-trigger-event-as-a-variable , `"${{ github.event.pull_request }}"` should work.

According to the GitHub docs, `GITHUB_HEAD_REF` only gets set for pull requests. But I should have used `-v` not '-z'.

Anyway, I've changed that to use `"${{ github.event.pull_request }}"

andy5995 replied:
> Anyway, I've changed that to use `"${{ github.event.pull_request }}"`

[Error, bad substitution](https://github.com/andy5995/appimage.github.io/runs/3001562669?check_suite_focus=true#step:4:1398)

I've changed it back to `GITHUB_HEAD_REF`

JulianGro replied:
You should be able to do something like:
test.yml
```bash
if [[ "${{ github.event.pull_request }}" = true ]]; then
          echo "IS_PULLREQUEST" >> $GITHUB_ENV
fi
```
and then

worker.sh
```bash
if [ $IS_PULLREQUEST != "false" ]; then
```

andy5995 replied:
> You should be able to do something like:
> test.yml
> 
> ```shell
> if [[ "${{ github.event.pull_request }}" = true ]]; then
>           echo "IS_PULLREQUEST=1" >> $GITHUB_ENV
> fi
> ```

I think I'm [doing it correctly](
https://docs.github.com/en/actions/reference/context-and-expression-syntax-for-github-actions#example-setting-an-environment-variable), but [I don't see that it's getting set to anything](https://github.com/AppImage/appimage.github.io/pull/2560/files#diff-faff1af3d8ff408964a57b2e475f69a6b7c7b71c9978cccc8f471798caac2c88R9).

JulianGro replied:
Yeah I would think that should work.
The code I suggested is from a working example: https://github.com/vircadia/vircadia/blob/master/.github/workflows/pr_build.yml

Maybe they had a reason to do it the way they did, but I would have too dig far too deep into the archives to justify looking it up.

JulianGro replied:
Just realized that we set the `env` before the actual job in our actions: https://github.com/vircadia/vircadia/blob/master/.github/workflows/pr_build.yml
You put the `env` "inside" the job. Maybe try moving it accordingly?

andy5995 commented on .github/workflows/test.yml:
```diff
@@ -0,0 +1,44 @@
+    runs-on: ubuntu-20.04
+    env:
+      DISPLAY: :99
+      IS_PULLREQUEST: ${{ github.event.pull_request }}
```
Here is where I'm trying to set it conditionally

andy5995 commented on .github/workflows/test.yml:
```diff
@@ -0,0 +1,43 @@
+
+env:
+  DISPLAY: :99
+  IS_PULLREQUEST: ${{ github.event.pull_request }}
```
When this is used, and I make a PR, I get `Error: .github/workflows/test.yml (Line: 6, Col: 19): A mapping was not expected`

https://github.com/andy5995/appimage.github.io/runs/3002416025?check_suite_focus=true#step:1:25

andy5995 replied:
When I make a PR, `${{ github.event.pull_request }}` is substituted for "Object"; if not a PR, the length of the string is zero.

https://github.com/andy5995/appimage.github.io/runs/3002504967?check_suite_focus=true

andy5995 commented on .github/workflows/test.yml:
```diff
@@ -0,0 +1,43 @@
+
+env:
+  DISPLAY: :99
+  IS_PULLREQUEST: ${{ github.event_name == 'pull_request' }}
```
This should work fine (returns "false")

https://github.com/andy5995/appimage.github.io/runs/3002651747?check_suite_focus=true#step:2:13

JulianGro commented on code/worker.sh:
```diff
@@ -533,31 +533,31 @@ echo "==========================================="
-git config --global user.email "travis@travis-ci.org"
-git config --global user.name "Travis CI"
+#git config --global user.email "travis@travis-ci.org"
+#git config --global user.name "Travis CI"
```
This should of course be cleaned up if it works.

JulianGro reviewed (approved):
Don't know if it works, but the code looks good to me.

probonopd commented:
Thanks @andy5995, this is super useful!

`TRAVIS_PULL_REQUEST` was used to determine whether we were running as part of a pull request. On GitHub Actions, we can probably use one of the [following](https://docs.github.com/en/actions/reference/environment-variables):

* `GITHUB_HEAD_REF`: Only set for pull request events. The name of the head branch.
* `GITHUB_BASE_REF`: Only set for pull request events. The name of the base branch.

I see that you are using one of those now, looks good.

> runs-on: ubuntu-latest

We need to run the test on the **oldest still-supported LTS release of Ubuntu**, which currently is **bionic**.
Possibly we need to use the Docker container from https://hub.docker.com/_/ubuntu?

mirkobrombin commented:
> Thanks @andy5995, this is super useful!
> 
> `TRAVIS_PULL_REQUEST` was used to determine whether we were running as part of a pull request. On GitHub Actions, we can probably use one of the [following](https://docs.github.com/en/actions/reference/environment-variables):
> 
> * `GITHUB_HEAD_REF`: Only set for pull request events. The name of the head branch.
> * `GITHUB_BASE_REF`: Only set for pull request events. The name of the base branch.
> 
> I see that you are using one of those now, looks good.
> 
> > runs-on: ubuntu-latest
> 
> We need to run the test on the **oldest still-supported LTS release of Ubuntu**, which currently is **bionic**.
> Possibly we need to use the Docker container from https://hub.docker.com/_/ubuntu?

Ubuntu 18.04 is still supported by GitHub Actions https://github.com/actions/virtual-environments#available-environments

JulianGro commented:
Yeah. They seem to generally keep support until non-enterprise support runs out.
They do have a lot of updated software on their VMs, but as far as I know all of that is build related and no "normal" dependencies are updated as well.

andy5995 commented:
The caching doesn't appear to be working. I'll have another look tomorrow. If anybody wants to make changes in the meantime, maintainers have write access.

JulianGro commented:
I don't think you really need to worry about caching apt packages. It's one of the two officially supported Linux systems. They will have repositories or caches in their datacenters.

probonopd commented:
Getting

![image](https://user-images.githubusercontent.com/2480569/125169015-d41e5100-e197-11eb-81e2-5fc257a44b57.png)

https://github.com/AppImage/appimage.github.io/runs/3036537333

Had this ever been tested successfully?

mirkobrombin commented:
`apt update` before installation should fix, it is not using external repositories

probonopd commented:
Makes me wonder if anyone had actually tested this...

andy5995 commented:
> Makes me wonder if anyone had actually tested this...

I suspect the problem is a temporary network issue. 

Yes, I tested it as much as I could on my repo with GitHub Actions. That step always passed, but the main test did not. I was expecting more feedback before it was merged.

![image](https://user-images.githubusercontent.com/16764864/125171123-1df73f00-e178-11eb-8f7a-97ca96f016ad.png)

https://github.com/andy5995/appimage.github.io/runs/3011839430?check_suite_focus=true

probonopd commented:
Now I am running into:

`Last changed files from the FILES variable:` contains ALL files, not just the one that was last changed.

andy5995 commented:
I noticed that too. I'm not sure why. I copied what was in the .travis.yml file with very few modifications. I added `sudo` before the apt-get and npm commands.

A couple changes related to FILES though:

I changed  `if [ x"$FILES" == x ]` to `if [ -z "$FILES" ]`

and if `[ x"$FILES" != x ]` to `if [ -n "$FILES" ]`
